### PR TITLE
Add testing for precise and non precise modes

### PR DESCRIPTION
**Pull request by Daniel4078**
finish individual task: 
https://trello.com/c/lRsLqmYg/1863-testing-for-precise-and-non-precise-modes-in-the-tests

### CODE REVIEW

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 1]
```diff
@@ -1876,7 +1899,9 @@ def get_ret_and_flattened_np_array(
         backend_to_test, fn, test_compile=test_compile, args=args, kwargs=kwargs
     )
     with update_backend(backend_to_test) as ivy_backend:
+        ivy.set_precise_mode(precision_mode)
         ret = fn(*args, **kwargs)
+        ivy.unset_precise_mode()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:1904]: We shouldn't call global `ivy` but instead `ivy_backend`

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 2]
```diff
@@ -1894,10 +1919,13 @@ def get_frontend_ret(
     frontend_fn,
     *args,
     as_ivy_arrays=True,
+    precision_mode=False,
     **kwargs,
 ):
     with update_backend(backend) as ivy_backend:
+        ivy.set_precise_mode(precision_mode)
         ret = frontend_fn(*args, **kwargs)
+        ivy.unset_precise_mode()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:1928]: We shouldn't call global `ivy` but instead `ivy_backend`

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 3]
```diff
@@ -933,13 +952,14 @@ def _gt_grad_fn(all_args):
                 kwargs=kwargs,
             )(*args, **kwargs)
             return gt_backend.nested_map(ret, gt_backend.mean, include_derived=True)
-
+        ivy.set_precise_mode(test_flags.precision_mode)
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:955]: We shouldn't call global `ivy` but instead `gt_backend`

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 4]
```diff
@@ -933,13 +952,14 @@ def _gt_grad_fn(all_args):
...
         _, grads_from_gt = gt_backend.execute_with_gradients(
             _gt_grad_fn,
             [args, kwargs, 1],
             xs_grad_idxs=xs_grad_idxs,
             ret_grad_idxs=ret_grad_idxs,
         )
+        ivy.unset_precise_mode()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:962]: We shouldn't call global `ivy` but instead `gt_backend`

**Review by CatB1t — CHANGES_REQUESTED** [review]

**Comment by Daniel4078** [comment]
I still need to add generation of the new precision_mode parameter in the test_flags object. Is there any hint or advices?

**Comment by Daniel4078** [comment]
It seems like dispite me adding the generation strategy for precision mode, it fails to be generated or not being passed correctly as shown here: E   TypeError: accept.<locals>.init_method_flags() missing 1 required keyword-only argument: 'precision_mode'.

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 5]
```diff
@@ -1975,7 +1998,9 @@ def get_frontend_ret(
             args, kwargs = ivy_backend.nested_map(
                 (args, kwargs), _frontend_array_to_ivy, include_derived={tuple: True}
             )
+        ivy_backend.set_precise_mode(precision_mode)
         ret = frontend_fn(*args, **kwargs)
+        ivy_backend.unset_precise_mode()
```
[the remark this answers is not in the document]
Comment (Daniel4078) [reply ivy_tests/test_ivy/helpers/function_testing.py:2003]: Can you give me a more detailed example of how to use the context manager? As I did not know that I added anything with the form of PreciseMode(precision_mode).

[the remark this answers is not in the document]
Comment (Daniel4078) [reply ivy_tests/test_ivy/helpers/function_testing.py:2022]: @sherry30 Currently I found there used to be another context manager implemented in ivy_tests/test_ivy/helpers/pipeline_helper.py as WithBackendContext. But I found that it is somehow removed from the current main branch, so I am not sure how to implement and use it. Can you give some hint?

**Comment by Daniel4078** [comment]
ivy-gardener

**Pull request merged**